### ISSUE TITLE
Multiple runs on same data with early_stopping not working in GPU training with dask_cudf

### ISSUE THREAD

**Issue by alexHeu** [dmlc/xgboost#8584]
Hi,

currently I am working on a multi-gpu workflow using dask_cudf together with xgboost 1.7.1 and early stopping.
The basic workflow works fine:

However, if I want to iterate multiple times over the same data, e.g. with different parameters, it appears that the early stopping is only working for the first iteration. In subsequent iteration training is not starting and it only prints the validation loss for node 0:
![image](https://user-images.githubusercontent.com/8157458/207072011-fb09e686-e6da-4c9e-8cce-f9ac85a5055e.png)
 
The basic workflow looks like this:
```python
cluster = LocalCUDACluster(n_workers=4)
client = Client(cluster)

df = dask_cudf.read_parquet("path_to_parquet")


X_train = df[df["CalDay"] < validation_start_date][features]
y_train = df[df["CalDay"] < validation_start_date][label]

X_val = df[df["CalDay"] >= validation_start_date][features]
y_val = df[df["CalDay"] >= validation_start_date][label]


def fit_xgboost(params, X_train, y_train, X_val, y_val):
  es = xgb.callback.EarlyStopping(rounds=early_stopping_rounds, save_best=False)
 
  Xy = dxgb.DaskQuantileDMatrix(client, X_train, y_train)
  Xy_valid = dxgb.DaskDMatrix(client, X_val, y_val)
 
  booster = xgb.dask.train(
    client,
    params,
    Xy,
    evals=[(Xy_valid, "Valid")],
    num_boost_round=1000,
    callbacks=[es],
    verbose_eval=True
  )
  return booster 

for params in params_list:
  fit_xgboost(params, X_train, y_train, X_val, y_val)
```

Thanks for any help that you can provide!

**Comment by trivialfis** [comment]
Could you please share the size of your dataset and the number of GPUs you are using? It seems dask is pulling all subsequent data into one worker.

**Comment by alexHeu** [comment]
@trivialfis Of course, currently I am on 4 V100 GPUs. I could try out machines with 2 or 1 GPU as well but havent so far.
The dataset that I am currently looking at has 3 152 629  rows and 131 columns.
I noticed this behavior with larger and smaller datasets as well.

For me it seems that there is some random element as well if it works or not, e.g. if I iterate over different subsets/groups of the same dataset and train a model per group then it works for some groups while it does not for others.

**Comment by trivialfis** [comment]
Have you tried to use `dask_ml` for splitting the data? https://github.com/dmlc/xgboost/blob/15a88ceef0f60d461925bb5133716ab67de6ca8e/demo/dask/dask_callbacks.py#L8

**Comment by trivialfis** [comment]
NVM, it seems you are doing some sort of time series task. Spiting data on a distributed system can be difficult sometime. I will take a deeper look later.

**Comment by trivialfis** [comment]
I think the issue is in data split, XGBoost doesn't move data and runs on the workers where data is resided on. If there's only one worker being used, then all the data is on that worker.

**Comment by alexHeu** [comment]
Yes it is a timeseries use-case.
I have around 30 datasets and on these I want to train around 200 models in total (1 per group). Therefore I do a lot of iterating over groups and/or parametersets and would like to use early stopping for each of these.

**Comment by trivialfis** [comment]
Could you please try to `repartition` the dataset after each iteration of tuning?

**Comment by alexHeu** [comment]
@trivialfis same issue unfortunately
